### PR TITLE
notifications: Integrate ionic notifications

### PR DESCRIPTION
**Pull request by DolceTriade**

### CODE REVIEW

File: backend/api_calls.py [hunk 1]
```diff
@@ -141,6 +141,7 @@ def Register(request, firebase):
   Args:
     userId: Unique userId added to the user list.
     name: A name to use to reference a player.
+    device_id: The device associated with the user. Used for notifications.
```
Comment (IsaacG) [inline backend/api_calls.py]: Are you using device_id in register?

File: backend/api_calls.py [hunk 2]
```diff
@@ -1055,6 +1056,29 @@ def MarkNotificationSeen(request, firebase):
...
 
+def RegisterUserDevice(request, firebase):
+  """Register a user device to a userId.
+
+  Validation:
+    userId must exist.
+
+  Args:
+    userId: User id to associate with.
+    deviceToken: Ionic device token.
```
Comment (IsaacG) [inline backend/api_calls.py:1092]: If only one device will be supported, would it make more sense to just do it in register()?

File: backend/api_calls.py [hunk 3]
```diff
@@ -1055,6 +1056,29 @@ def MarkNotificationSeen(request, firebase):
...
+
+  Firebase entries:
+    /users/%(userId)/deviceToken
+"""
+  valid_args = ['userId']
+  required_args = list(valid_args)
+  required_args.extend(['deviceToken'])
+  ValidateInputs(request, firebase, required_args, valid_args)
+  put_data = {
```
Comment (IsaacG) [inline backend/api_calls.py]: Stylistic personal preference: collapse the put_data assignment to one line.

File: backend/ionic.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    self.app_id = app_id
+    self.access_token = 'Bearer %s' % access_token
+    self.security_tag = security_tag
+    self.Test()
+
+  def GetAuthHeaders(self):
+    """Generate authoriztaion headers."""
+    return {
+      'Authorization': self.access_token,
+      'Content-Type': 'application/json',
```
Comment (IsaacG) [inline backend/ionic.py]: I'm pretty sure the requests module puts this header in for you when you send JSON data.
Also, does requests require updating the requirements.txt?

Comment (IsaacG) [reply backend/ionic.py]: Plus, it's not an auth-related header

File: backend/ionic.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  def Test(self):
+    """Validate credentials.
+
+    Returns:
+      None. Raises exception on error.
+    """
+    r = requests.get(BuildUrl('auth/test'), headers=self.GetAuthHeaders())
+    r.raise_for_status()
+
+  def Post(self, url_suffix, json_data):
```
Comment (IsaacG) [inline backend/ionic.py]: Technically, it's just 'data' and the implementation details convert it to JSON

File: backend/ionic.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    """Post data to ionic."""
+    r = requests.post(BuildUrl(url_suffix), json=json_data, headers=self.GetAuthHeaders())
+    r.raise_for_status()
+    return r
+
+  def SendNotification(self, tokens, message, destination):
+    """Send a notification to the device tokens specified.
+
+    Args:
+      tokens: An array of device tokens to send to.
```
Comment (IsaacG) [inline backend/ionic.py]: Rename to "devices"?

File: backend/ionic.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      message: Message to send in the notification.
+      destination: Destination the notificatoin should open.
```
Comment (IsaacG) [inline backend/ionic.py]: Typo in notification.
Is this a URL?

File: backend/ionic.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        'payload': {
+          'destination': destination,
+        }
+      },
+      'profile': self.security_tag
+    }
+    if tokens:
+      data['tokens'] = tokens
+    else:
+      data['send_to_all'] = True
```
Comment (IsaacG) [inline backend/ionic.py]: Do we ever want to do this? If not, might be better to not have it at all so when we accidentally set tokens=None we don't spam people.

File: backend/notifications.py [hunk 9]
```diff
@@ -22,6 +24,23 @@ def HandleNotification(firebase, notification_id, notification):
     firebase.put('/players/%s/notifications' % player,
                   notification_id, notification)
 
+  if 'app' in notification:
+    player_set = set(players)
```
Comment (IsaacG) [inline backend/notifications.py]: May as well just make players a set in the first place.
Avoid placing data types in var names.

File: backend/notifications.py [hunk 10]
```diff
@@ -22,6 +24,23 @@ def HandleNotification(firebase, notification_id, notification):
...
+    # This is really depressing... This is only barely acceptable because we
+    # expect less than 100 players for the minigame.
+    users = firebase.get('/users', None)
+    tokens = set()
+    if users is None:
+      logging.error('Error querying all users.')
+      return
+    for user_id, user in users.iteritems():
```
Comment (IsaacG) [inline backend/notifications.py]: If you're not using user_id, don't generate it.
for user in users.values():

File: backend/secrets.py [hunk 11]
```diff
@@ -2,3 +2,6 @@
 # https://console.firebase.google.com/project/trogdors-29fa4/settings/serviceaccounts/databasesecrets
 FIREBASE_SECRET = 'd0h365xNGuJU7A3rB65bAJ4v59VMLJiaNY5R0D6C'
 FIREBASE_EMAIL = ''
+IONIC_TOKEN = ''
+IONIC_APPID = ''
+IONIC_SECURITY_TAG = ''
```
Comment (IsaacG) [inline backend/secrets.py]: These values should be shoved on Valentine or something

File: backend/ionic.py (hunk 4 above)
Comment (DolceTriade) [reply backend/ionic.py]: Yeah, I was having issues earlier so I added it. I'm not sure if this is what fixed it, but I'll clarify. Also, requests is already in requirements.txt.

File: backend/api_calls.py (hunk 1 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Whoops. Leftover...

File: backend/api_calls.py (hunk 2 above)
Comment (DolceTriade) [reply backend/api_calls.py:1092]: We don't allow updating user data so I made a separate method so you can overwrite the old value.

File: backend/api_calls.py (hunk 3 above)
Comment (DolceTriade) [reply backend/api_calls.py]: Done.

File: backend/ionic.py (hunk 5 above)
Comment (DolceTriade) [reply backend/ionic.py]: Done.

File: backend/ionic.py (hunk 6 above)
Comment (DolceTriade) [reply backend/ionic.py]: Done.

File: backend/ionic.py (hunk 7 above)
Comment (DolceTriade) [reply backend/ionic.py]: Done. Yes, it's a URL for now.

File: backend/ionic.py (hunk 8 above)
Comment (DolceTriade) [reply backend/ionic.py]: Removed.

File: backend/notifications.py (hunk 10 above)
Comment (DolceTriade) [reply backend/notifications.py]: Done

File: backend/notifications.py (hunk 9 above)
Comment (DolceTriade) [reply backend/notifications.py]: Done

File: backend/secrets.py (hunk 11 above)
Comment (DolceTriade) [reply backend/secrets.py]: Yes. I still need to talk to you on how these values get set when you deploy to prod.

File: backend/api_calls.py (hunk 2 above)
Comment (IsaacG) [reply backend/api_calls.py:1092]: Why would we not have a method that updates user data? I mean, that's exactly what this method is doing, right? This method is essentially an UpdateUser that only supports on entry in the user data.

File: backend/secrets.py (hunk 11 above)
Comment (IsaacG) [reply backend/secrets.py]: Might be worth having a simple shell script that does the deploy.
First time it's ran, we can prompt for the secrets and auto-create a secrets.py in some ~/.config path then on deploy, run the pip command, copy the file in and run the gcloud deploy.

File: backend/api_calls.py (hunk 2 above)
Comment (DolceTriade) [reply backend/api_calls.py:1092]: Correct, but I assumed that the lack of an UpdateUser method was a design decision.

Comment (IsaacG) [reply backend/api_calls.py:1092]: Nah. There just wasn't anything that made sense to update.
Given that you *are* writing a method which updates the user data, is there any reason not to include the device in the original register vs requiring a register followed by a second RPC to update that data with the device?

Comment (DolceTriade) [reply backend/api_calls.py:1092]: The main reason is that not everyone is likely to call /register from their device. They may do it from the website and then install the app. I think this should be separate and should be called everytime someone logs into an app via mobile. Whether this is part of the larger /updateUser call or specifically with a /registerUserDevice call. In this case, I would argue that a call specifically for this makes more sense given that we are specifically interested in registering a device for the user and not necessarily updating any other user related metadata.

Comment (IsaacG) [reply backend/api_calls.py:1092]: That sounds reasonable.

If we do end up adding a "UpdateUser", we can change this method to call UpdateUser() and pass in just the device token.

**Pull request merged**